### PR TITLE
Refactor Project to Crate

### LINKED ISSUE #113: Refactoring Project --> Crate
We adopt 'crate' for the bundles of GPR project + Ada code downloaded by alr. This was previously called Project through alire, with the risk of confusion with GPR projects or abstract projects.

This entails:
1. Plain refactoring of Project --> Crate through code
2. A possible opportunity to enhance `Alire.Projects` to better store all crate-exclusive information which cannot be overriden by releases.

### PR DESCRIPTION
**Pull request by mosteo**
Fixes #113.

### CODE REVIEW

File: src/alire/alire-containers.ads [hunk 1]
```diff
@@ -31,12 +31,12 @@ package Alire.Containers with Preelaborate is
                                              Releases."=");
    subtype Release_H is Release_Holders.Holder;
 
-   package Project_Release_Maps is new Ada.Containers.Indefinite_Ordered_Maps
-     (Alire.Project, Releases.Release, "<", Releases."=");
-   type Release_Map is new Project_Release_Maps.Map with null record;
+   package Name_Release_Maps is new Ada.Containers.Indefinite_Ordered_Maps
```
Comment (Fabien-Chouteau) [inline src/alire/alire-containers.ads]: Shouldn't this be `Crate_Release_Maps`?

File: src/alire/alire.ads [hunk 2]
```diff
@@ -51,8 +51,8 @@ package Alire with Preelaborate is
    function "+" (S : UString) return String
    renames UStrings.To_String;
 
-   subtype Project_Character is Character
-      with Static_Predicate => Project_Character in
+   subtype Crate_Character is Character
+      with Static_Predicate => Crate_Character in
          'a' .. 'z' | '0' .. '9' | '_' | Extension_Separator;
```
Comment (Fabien-Chouteau) [inline src/alire/alire.ads:56]: Completely unrelated: Why hyphen is not allowed in crate name?

File: src/alr/alr-commands-search.adb [hunk 3]
```diff
@@ -104,14 +104,14 @@ package body Alr.Commands.Search is
...
-           (Crate : Alire.Projects.With_Releases.Crate) is
+           (Crate : Alire.Crates.With_Releases.Crate) is
          begin
             if Cmd.Full then
                for Release of Crate.Releases loop
                   List_Release (Release);
                   Busy.Step;
                end loop;
-            else
+            elsif not Crate.Releases.Is_Empty then
```
Comment (Fabien-Chouteau) [inline src/alr/alr-commands-search.adb:114]: This is more than renaming. What is the reason for the change?

File: src/alr/alr-paths.ads [hunk 4]
```diff
@@ -92,7 +92,7 @@ private
    function Alr_Source_Folder return String
    is (OS_Lib.Getenv (Alire.Environment.Source, Alr_Config_Folder / "alire"));
 
-   function Projects_Folder      return String
-   is (Alr_Working_Cache_Folder / "projects");
+   function Releases_Folder      return String
+   is (Alr_Working_Cache_Folder / "releases");
```
Comment (Fabien-Chouteau) [inline src/alr/alr-paths.ads]: Why not `"crates"` here?

File: src/alire/alire-containers.ads (hunk 1 above)
Comment (mosteo) [reply src/alire/alire-containers.ads]: Sometimes I have difficulties choosing between name and crate depending on context. I change it if you find it more consistent.

File: src/alire/alire.ads (hunk 2 above)
Comment (mosteo) [reply src/alire/alire.ads:56]: When the index was Ada-based I tried to keep things more closer to Ada syntax (indeed the hyphens in the toml syntax ruffle me the wrong way, but since it's a matter of taste...) So no actual reason now I'd say.

File: src/alr/alr-commands-search.adb (hunk 3 above)
Comment (mosteo) [reply src/alr/alr-commands-search.adb:114]: Ah, good eye. It's a bug fix needed after #293, for the case where we had a crate without releases (there aren't in the current index). This PR was initially on top of my externals experiments in #295. I caught the bug then, and when rebasing this PR I didn't realize this was a functional change (even if it's a bug fix).

File: src/alr/alr-paths.ads (hunk 4 above)
Comment (mosteo) [reply src/alr/alr-paths.ads]: I've realized that many times when we talk about crates we are talking about a particular release of a crate. In this case, we are downloading releases, so that's it. Although it's kind of splitting hairs probably.

Comment (mosteo) [reply src/alr/alr-paths.ads]: On further thought, this should probably be better "dependencies"

File: src/alire/alire.ads (hunk 2 above)
Comment (Fabien-Chouteau) [reply src/alire/alire.ads:56]: I'm mostly curious, I do like hyphens in project names better than underscores (`lvgl-ada` vs `lvgl_ada`) but since allowing both would be a source of confusion let's keep it how it is.

File: src/alire/alire-containers.ads (hunk 1 above)
Comment (Fabien-Chouteau) [reply src/alire/alire-containers.ads]: From what I know of the code it makes more sense to me.

Comment (mosteo) [reply src/alire/alire-containers.ads]: Done in e07df2a

File: src/alr/alr-paths.ads (hunk 4 above)
Comment (mosteo) [reply src/alr/alr-paths.ads]: Done in e07df2a

**Review by Fabien-Chouteau — APPROVED** [review]
